processor/otel/metrics.go

```diff
@@ -89,6 +90,91 @@ func (c *Consumer) convertResourceMetrics(resourceMetrics pdata.ResourceMetrics,
 	}
 }
 
+// This builder groups system metrics with the same name in a map entry, allowing to perform aggregation
+// Otel specification : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/system-metrics.md
+// ex : total memory = system.memory.usage (state = free) + system.memory.usage (state = used)
+type apmMetricBuilder struct {
+	metricList []pdata.Metric
+	metricName string
+}
+
+func (b *apmMetricBuilder) build(ms metricsets) {
```

Comment:
Do you think we could simplify this by splitting the aggregation and emitting of Elastic metrics? I'd also like it if we could refactor the existing code so we do all translation of OTel -> Elastic metrics in the same way. That probably means moving the switch out of metricsets.upsert.

Finally, I think it might be a bit more straightforward if we made apmMetricBuilder less generic. I'm imagining something like this:

```go
type apmMetricsBuilder struct {
    // System metrics
    cpuCount int // from system.cpu.utilization's cpu attribute
    nonIdleCPUUtilizationSum float64
    freeMemoryBytes int64
    usedMemoryBytes int64

    // JVM metrics
    jvmMemoryArea int64
    jvmGCTime map[string]int64
    jvmGCCount map[string]int64
    jvmMemory map[jvmMemoryKey]int64
}

type jvmMemoryKey struct {
    area string
    type_ string
    pool string // will be "" for non-pool specific memory metrics
}

// accumulate processes m, translating to and accumulating equivalent Elastic APM metrics in b.
func (b *apmMetricsBuilder) accumulate(m pdata.Metric) {
}

// emit upserts Elastic APM metrics into ms from information accumulated in b.
func (b *apmMetricsBuilder) emit(ms metricsets) {
}
```

You would iterate through all OTel metrics, upserting the original OTel metric and calling `apmMetricsBuilder.accumulate`. Then finally, call `apmMetricsBuilder.emit` to produce `system.memory.total` and `system.cpu.total.norm.pct`, as well as `runtime.jvm.memory.area` etc.

WDYT?

Reply:
This is a fantastic idea ! I will try to propose something that matches your vision in my next commit.